### ISSUE TITLE
Firmware Update Timeout (30seconds) too low for some deices on a busy network

### ISSUE THREAD

**Issue by warthog9** [AlCalzone/node-zwave-js#4407]
### Is your problem within Home Assistant (Core or Z-Wave JS Integration)?

NO, my problem is NOT within Home Assistant or the ZWave JS integration

### Is your problem within ZWaveJS2MQTT?

NO, my problem is NOT within ZWaveJS2MQTT

### Checklist

- [X] I have checked the [troubleshooting section](https://zwave-js.github.io/node-zwave-js/#/troubleshooting/index) and my problem is **not** described there.

- [X] I have read the [changelog](https://github.com/zwave-js/node-zwave-js/blob/master/CHANGELOG.md) and my problem was **not** mentioned there.

### Describe the bug

What causes the bug?

Attempting to do a firmware update on an Inovelli LZW60

What do you observe?

It starts sending packets (initiated from zwavejs2mqtt) runs for a bit fine, and then the firmware update hits, cleans up and a few seconds later gets a firmware request but zwave-js doesn't think that there's an ongoing firmware update and force aborts it.

What did you expect to happen?

Ideally it not time out

  1. Initiate update from zwavejs2mqtt
  2. Wait
  3. Firmware update times out

log entries:

2022-03-28T04:01:34.309Z CNTRLR   [Node 188] Firmware update timed out
2022-03-28T04:01:40.367Z CNTRLR « [Node 188] Received Firmware Update Get, but no firmware update is in progress
                                  . Forcing the node to abort...
                                  
I'd propose changing the timeout (30000ms or 30seconds) to at least 1minute, but more likely 5minutes, as 30 seconds seems to be just tight enough that the update can't complete, and I'm sure the LZW60 isn't the only device that may run into this.  In my own testing changing it to 20minutes (yes I recognize that's a smidge ridiculous) seems to resolve the issue

### Device information

Manufacturer: Inovelli
Model name: LZW60
Node ID in your network: 188

### How are you using `node-zwave-js`?

- [ ] `zwavejs2mqtt` Docker image (latest)
- [ ] `zwavejs2mqtt` Docker image (dev)
- [ ] `zwavejs2mqtt` Docker manually built _(please specify branches)_
- [ ] `ioBroker.zwave2` adapter _(please specify version)_
- [ ] `HomeAssistant zwave_js` integration  _(please specify version)_
- [ ] `pkg`
- [ ] `node-red-contrib-zwave-js` _(please specify version, double click node to find out)_
- [ ] Manually built from GitHub _(please specify branch)_
- [X] Other _(please describe)_

### Which branches or versions?

version:
`node-zwave-js` branch: whatever zwavejs2mqtt pulls in, the timeout is identical in master when I looked
`zwavejs2mqtt` branch: master

### Did you change anything?

yes (please describe)

### If yes, what did you change?

I made a change to node_modules/zwave-js/build/lib/node/Node.js to change the timeout to timeoutFirmwareUpdate from 30000 to  1.2e6 ms to test this, otherwise this is basically unchanged

### Did this work before?

Don't know, this is a new device

### Attach Driver Logfile

[log_output.txt](https://github.com/zwave-js/node-zwave-js/files/8359742/log_output.txt)

**Comment by warthog9** [comment]
I'll note the log file is just the relevant portion of the firmware update for node 188, the rest of the traffic is somewhat irrelevant to this discussion

**Comment by skynet01** [comment]
I had similar issue when updating firmware on LZW31-SN (firmware was around 285kb). My network is 50 devices. I had to try the process twice, the only way I was able to get the firmware through is minimize the network traffic (tell people not to move, turn on lights etc). Hopefully time out can fix this :)

**Comment by AlCalzone** [comment]
We can definitely investigate increasing the timeout to help with this.

This does sound a bit like intermittent connectivity issues of the switch though. The log doesn't contain any relevant details, since it's not on loglevel debug, but the delay between fragments 2 and 3 is already cutting it very close:
```
2022-03-28T03:56:25.875Z CNTRLR » [Node 188] Sending firmware fragment 2 / 4981
2022-03-28T03:56:49.945Z CNTRLR » [Node 188] Sending firmware fragment 3 / 4981
```
It could also be that the controller is waiting for something else here, not sure.

Do note that a firmware update is mostly controlled by the device that gets updated, which individually requests each firmware fragment. Z-Wave JS just considers the process to be stuck or aborted when it doesn't hear from the device for a while. The fact that it (probably) sometimes takes >20s to send a message to the controller is a bit bothersome IMO.

Have you used the mesh health check tools in zwavejs2mqtt yet? They can be found on the network map.acting up.

**Comment by warthog9** [comment]
Quick check of the nodes (now that they are re-flashed and back in the network) looks like they are healthy and fine. (6,7,8,8,7 scores) so I don't think it was specifically an issue with the general health of the network (minor update)

**Comment by AlCalzone** [comment]
Scores of 6 or 7 might already cut it a bit close - low latency but possibly not a strong link:
![grafik](https://user-images.githubusercontent.com/17641229/162960999-2dbe1491-4060-4b22-be57-ee2c7892b8b6.png)

What are the min. powerlevel and SNR margins?

**Comment by twasilczyk** [comment]
I have the exact same problem and I see a number of other threads about 30s timeout being too short, e.g.:
- https://community.home-assistant.io/t/a-number-of-problems-including-bugs-break-firmware-update/564833 (this one's mine)
- https://github.com/zwave-js/node-zwave-js/issues/3094
- https://github.com/zwave-js/node-zwave-js/issues/5646

This doesn't seem to be affected by network traffic (I purposedly removed sensors from my network for the test), but it does seem affected by connection quality and number of nodes (upgrading 500 stick to 800 stick helped quite a lot, but not completely).

Can we go forward with extending that 30s timeout? If the update process takes 10s of minutes, what's the point of aborting after 30s of no successful packet transmission if the other side still tries to continue it (verified both in my logs and one of the bugs listed above). It would be good to somehow measure how long is update timeout on the device side, but at least Aeotec and Zooz have it greater than 30s. I would suggest 5 minutes as warthog9 suggested.

To AlCalzone point, it's not that it takes >20s to send a message to the controller, but it sometimes gets stuck for a few retries and each retry takes a couple seconds. If this happens often, with a very long process like a FW update there's a good chance to hit the timeout at some point. Since AlCalzone asked for more detailed logs, I'm pasting mine: [zwavejs_current.zip](https://github.com/zwave-js/node-zwave-js/files/11680853/zwavejs_current.zip)

**TL;DR: Please increase FW update timeout to 1~5 minutes :)**

**Comment by AlCalzone** [comment]
Actually, it does sometimes take in the order of 10-20 seconds to send a single command:
```
2023-06-07T18:16:56.234Z CNTRLR » [Node 019] Sending firmware fragment 2359 / 6055
2023-06-07T18:16:56.277Z DRIVER » [Node 019] [REQ] [SendDataBridge]
                                  │ source node id:   1
                                  │ transmit options: 0x25
                                  │ callback id:      100
                                  └─[Security2CCMessageEncapsulation]
                                    │ sequence number: 68
                                    └─[FirmwareUpdateMetaDataCCReport]
                                        report #: 2359
                                        is last:  false
2023-06-07T18:16:56.303Z DRIVER « [RES] [SendDataBridge]
                                    was sent: true
...
2023-06-07T18:17:13.067Z DRIVER « [REQ] [SendDataBridge]
                                    callback id:            100
                                    transmit status:        OK, took 16780 ms
                                    repeater node IDs:      5, 20
                                    routing attempts:       18
                                    protocol & route speed: Z-Wave, 100 kbit/s
                                    ACK RSSI:               -93 dBm
                                    ACK RSSI on repeaters:  -86 dBm, -99 dBm
                                    ACK channel no.:        0
                                    TX channel no.:         0
                                    route failed here:      17 -> 19
```
almost 17 seconds, and the next:
```
2023-06-07T18:17:13.106Z CNTRLR » [Node 019] Sending firmware fragment 2360 / 6055
2023-06-07T18:17:13.163Z DRIVER » [Node 019] [REQ] [SendDataBridge]
                                  │ source node id:   1
                                  │ transmit options: 0x25
                                  │ callback id:      101
                                  └─[Security2CCMessageEncapsulation]
                                    │ sequence number: 69
                                    └─[FirmwareUpdateMetaDataCCReport]
                                        report #: 2360
                                        is last:  false
2023-06-07T18:17:13.192Z DRIVER « [RES] [SendDataBridge]
                                    was sent: true
...
2023-06-07T18:17:33.530Z DRIVER « [REQ] [SendDataBridge]
                                    callback id:            101
                                    transmit status:        OK, took 20070 ms
                                    routing attempts:       25
                                    protocol & route speed: Z-Wave, 100 kbit/s
                                    ACK RSSI:               -97 dBm
                                    ACK channel no.:        0
                                    TX channel no.:         0
                                    route failed here:      12 -> 19
```
20 seconds.

Like in the previous report, this is an issue with the device connectivity. If it takes 20 seconds to send fragments to the device, I don't think you'll want to wait for all 6000+ fragments to be done at that rate - that's over 30 hours!

That said - we can increase the timeout but this **really** points at underlying issues with your mesh.

**Comment by twasilczyk** [comment]
I don't think it ever takes 20 seconds to send each fragment to the device - it's rather:
1. device downloads first 100-500 fragments smoothly
2. device tries to send FirmwareUpdateMetaDataCCGet, but the payload gets mangled
3. device tries again after a few seconds
4. steps 2&3 may repeat a few times
5. device continues to download next 500-ish fragments smoothly

If steps 2&3 repeat 5-10 times, it will easily hit the 30s timeout. You can see the device indeed was trying multiple times in step 2, because the sequence numbers between the last successful and the first successful get after *Dropping message with invalid payload* jumps a few numbers. And if increasing the timeout indeed pushed someone's update to 30 hours, they could still be able to abort the process.

Having said that, I do realize there's some problem with my (and surely others) mesh, but I'm having a very hard time tracking it down. Explained [it here](https://community.home-assistant.io/t/a-number-of-problems-including-bugs-break-firmware-update/564833/2).

So yes, please increase the timeout and I'll continue trying to fix my network.

**Comment by AlCalzone** [comment]
I literally quoted a part of the log where it took 20 seconds to reach the device 🤷🏻‍♂️

**Comment by twasilczyk** [comment]
Sorry, as I'm re-reading what I wrote I might not express this clearly: I don't think it ever takes 20 seconds to send every fragment, not a few odd ones (and the rest takes hundreds of milliseconds). The rest of my response doesn't deny that sometimes it takes 30+, but rather tries to explain why it happens.

EDIT: I also realized that in this specific case it actually took 17 and 20s to get a single message through. So yeah, both types of failures happen. I have no idea why the mesh picked nodes 17 and 12 to get the message through - these are pretty far from node 19.

**Comment by twasilczyk** [comment]
The change was finally pushed to zwave-js-ui - it works! Thank you very much, I was finally able to update my entire network without much fuss.
